pkg/audit/reverify_test.go

```diff
@@ -0,0 +1,423 @@
+
+func TestReverifyFailMissingShare(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount: 1, StorageNodeCount: 6, UplinkCount: 1,
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+
+		// - uploads random data
+		// - uses the cursor to get a stripe
+		// - creates one pending audit for a node holding a piece for that stripe
+		// - the actual share is downloaded to make sure ExpectedShareHash is correct
+		// - delete piece from node
+		// - calls reverify on that same stripe
+		// - expects one storage node to be marked as a fail in the audit report
+
+		err := planet.Satellites[0].Audit.Service.Close()
+		require.NoError(t, err)
+
+		ul := planet.Uplinks[0]
+		testData := make([]byte, 1*memory.MiB)
+		_, err = rand.Read(testData)
+		require.NoError(t, err)
+
+		err = ul.Upload(ctx, planet.Satellites[0], "testbucket", "test/path", testData)
+		require.NoError(t, err)
+
+		metainfo := planet.Satellites[0].Metainfo.Service
+		cursor := audit.NewCursor(metainfo)
+
+		var stripe *audit.Stripe
+		stripe, _, err = cursor.NextStripe(ctx)
+		require.NoError(t, err)
+		require.NotNil(t, stripe)
+
+		orders := planet.Satellites[0].Orders.Service
+		containment := planet.Satellites[0].DB.Containment()
+
+		verifier := audit.NewVerifier(zap.L(),
+			planet.Satellites[0].Transport,
+			planet.Satellites[0].Overlay.Service,
+			containment,
+			orders,
+			planet.Satellites[0].Identity,
+			128*memory.B,
+			5*time.Second)
+		require.NotNil(t, verifier)
+
+		projects, err := planet.Satellites[0].DB.Console().Projects().GetAll(ctx)
+		require.NoError(t, err)
+
+		bucketID := []byte(storj.JoinPaths(projects[0].ID.String(), "testbucket"))
+		shareSize := stripe.Segment.GetRemote().GetRedundancy().GetErasureShareSize()
+
+		pieces := stripe.Segment.GetRemote().GetRemotePieces()
+		rootPieceID := stripe.Segment.GetRemote().RootPieceId
+		limit, err := orders.CreateAuditOrderLimit(ctx, planet.Satellites[0].Identity.PeerIdentity(), bucketID, pieces[0].NodeId, rootPieceID, shareSize)
+		require.NoError(t, err)
+
+		share, err := verifier.GetShare(ctx, limit, stripe.Index, shareSize, int(pieces[0].PieceNum))
+		require.NoError(t, err)
+
+		pending := &audit.PendingAudit{
+			NodeID:            pieces[0].NodeId,
+			PieceID:           rootPieceID,
+			StripeIndex:       stripe.Index,
+			ShareSize:         shareSize,
+			ExpectedShareHash: pkcrypto.SHA256Hash(share.Data),
+			ReverifyCount:     0,
+		}
+
+		err = containment.IncrementPending(ctx, pending)
+		require.NoError(t, err)
+
+		// delete the piece from the first node
+		nodeID := stripe.Segment.GetRemote().GetRemotePieces()[0].NodeId
+		pieceID := stripe.Segment.GetRemote().RootPieceId.Derive(nodeID)
+		node := getStorageNode(planet, nodeID)
+		err = node.Storage2.Store.Delete(ctx, planet.Satellites[0].ID(), pieceID)
+		require.NoError(t, err)
+
+		report, err := verifier.Reverify(ctx, stripe)
+		require.NoError(t, err)
+
+		require.Len(t, report.Successes, 0)
```

Comment:
maybe we should make the other tests consistent with this one by testing the lengths of `report.Successes`, `report.Offlines`, `report.PendingAudits`, `report.Fails`. In `TestReverifySuccess` we are only testing one of those. This comment applies to more than just `TestReverifySuccess`

Reply:
that's a good idea!

Reply:
it deserves an issue to remember to do so!